apm-sniffer/apm-sdk-plugin/httpasyncclient-4.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/httpasyncclient/v4/HttpAsyncRequestExecutorInterceptor.java

```diff
@@ -48,12 +48,14 @@
     @Override
     public void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes, MethodInterceptResult result) throws Throwable {
         HttpContext context = CONTEXT_LOCAL.get();
+        Boolean isOutNotExit = CONTEXT_LOCAL_NOT_EXIT.get();
```

Comment:
Why need this? Is this method not in the main context thread?

Reply:
`ContextManager.activeSpan().isExit()` should be available if this is in the main thread, and`CONTEXT_LOCAL_NOT_EXIT` is not required.

Reply:
Because there is creating exit span in there, so it makes sense only existing in the main thread. You need to make the change very clear, otherwise, hard to review.

Reply:
it is not in the main context thread，flag isOutNotExit used to determine if the main thread is not exit span. The child thread only creates an exit span when the main thread is a non-exit span.

Reply:
> `ContextManager.activeSpan().isExit()` should be available if this is in the main thread, and`CONTEXT_LOCAL_NOT_EXIT` is not required.

In the main thread, the ContextManager.activeSpan() span in some scenarios is an Entry span, not an exit span, such as the example in the test  httpasyncclint scenario.

Reply:
I think you still miss the context. The whole scenario is `ES exit span`->`http exit span`,then only ES span left. 

`ContextManager.activeSpan()` could be
1. No exist, because http client is the first one.
1. Local span or entry span, because some other span exists.
1. Exit span exists, such as ES plugin is working.

So what? We are talking about the async HTTP client logic.  
First, the HTTP async span should start in the main thread, no matter what old codes look like. Then change into the `async` mode, and propagate the span into sub-thread, and tag, log or anything you need to do.
If there is a callback for HTTP client, you could choose to create a new segment and local span or just simply tag because such as it is invoked inside ES client.

Reply:
It seems that I have a problem before, the calculated span time should be wrong, and does not include the execution time of httpasycclient. Maybe I understand what you mean, I will try again.

There is a problem, in the asynchronous case, you need to create an exit span in the main thread, trigger prepareForAsync. When ContextManager.activeSpan() is Local span or entry span, you need to create create a new segment and local span in the child thread. At this time, you find that there is a duplicate creation of the exit span in the code, in the main thread. One, then the callback will appear again; don't know if this understanding is correct?

Reply:
Sure. Take your time. You only could have only one exit span with the accurate duration, and tags.